src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java

```
@@ -168,26 +212,72 @@ protected final void decodeSimpleTypes(JsonObject json, String name, StructuredR
 
     switch (schema.getType()) {
       case NULL:
-        json.add(name, JsonNull.INSTANCE); // nothing much to do here.
+        writer.nullValue(); // nothing much to do here.
         break;
       case INT:
       case LONG:
       case FLOAT:
       case DOUBLE:
-        json.addProperty(name, (Number) object);
+        writer.value((Number) object);
         break;
       case BOOLEAN:
-        json.addProperty(name, (Boolean) object);
+        writer.value((Boolean) object);
         break;
       case STRING:
-        json.addProperty(name, object.toString());
+        writer.value(object.toString());
         break;
+      // TODO CDAP-15256 write byte type to json writer
       default:
-        throw new IllegalStateException(
-          String.format("Field '%s' is of unsupported type '%s'", name, schema.getType()));
+        throw new IllegalStateException(String.format("Field '%s' is of unsupported type '%s'",
+                                                      name, schema.getType()));
     }
   }
 
+  private void decodeArray(JsonWriter writer, String name, Object value, Schema fieldSchema) throws IOException {
+    if (value == null) {
+      writer.nullValue();
+      return;
+    }
+
+    if (!(value instanceof Collection) && !value.getClass().isArray()) {
+      throw new IllegalArgumentException("The value should be of type collection or array. Got: " + value.getClass());
+    }
+
+    Schema componentSchema = BigQueryUtil.getNonNullableSchema(fieldSchema.getComponentSchema());
+    // Arrays within arrays are not allowed in big query
+    if (componentSchema.getType() == Schema.Type.ARRAY) {
+      throw new IllegalArgumentException("Nested Arrays is not a valid type in big query.");
+    }
+
+    writer.name(name);
+    writer.beginArray();
+    if (value instanceof Collection) {
+      for (Object element : (Collection) value) {
+        // big query does not allow null values in array items
+        if (element != null) {
```

this will silently drop null elements, which seems bad because it's modifying the input data. I think this should cause a failure, with an error message describing which field is causing the problem.

I also wonder if we should be checking the schema to make sure it's not an array of nullable elements, as that would be a deploy time failure instead of runtime.

---

Done, also Filed https://issues.cask.co/browse/CDAP-15345